src/ORM/DataList.php

```diff
@@ -1128,41 +1136,59 @@ private function fetchEagerLoadHasOne(
                 $hasOneID = $parentData[$hasOneIDField];
-                $fetchedIDs[] = $hasOneID;
-                $addTo[$hasOneID][] = $parentData['ID'];
+
+                if ($hasOneID > 0) {
+                    $hasOneClass = $schema->baseDataClass($parentData[$hasOneClassField] ?? $relationDataClass);
+
+                    $fetchedIDs[$hasOneClass][$hasOneID] = $hasOneID;
+                    $addTo[$hasOneClass][$hasOneID][] = $parentData['ID'];
+                }
             } elseif ($parentData instanceof DataObject) {
                 // $parentData represents another has_one record
                 $hasOneID = $parentData->$hasOneIDField;
-                $fetchedIDs[] = $hasOneID;
-                $addTo[$hasOneID][] = $parentData;
+
+                if ($hasOneID > 0) {
+                    $hasOneClass = $schema->baseDataClass($parentData->$hasOneClassField ?? $relationDataClass);
+
+                    $fetchedIDs[$hasOneClass][$hasOneID] = $hasOneID;
+                    $addTo[$hasOneClass][$hasOneID][] = $parentData;
+                }
             } elseif ($parentData instanceof EagerLoadedList) {
                 // $parentData represents a has_many or many_many relation
                 foreach ($parentData->getRows() as $parentRow) {
                     $hasOneID = $parentRow[$hasOneIDField];
-                    $fetchedIDs[] = $hasOneID;
-                    $addTo[$hasOneID][] = ['ID' => $parentRow['ID'], 'list' => $parentData];
+
+                    if ($hasOneID > 0) {
+                        $hasOneClass = $schema->baseDataClass($parentRow[$hasOneClassField] ?? $relationDataClass);
+
+                        $fetchedIDs[$hasOneClass][$hasOneID] = $hasOneID;
+                        $addTo[$hasOneClass][$hasOneID][] = ['ID' => $parentRow['ID'], 'list' => $parentData];
+                    }
                 }
             } else {
                 throw new LogicException("Invalid parent for eager loading $relationType relation $relationName");
             }
         }
 
-        $fetchedRecords = DataObject::get($relationDataClass)->byIDs($fetchedIDs)->toArray();
+        $fetchedRecords = [];
 
-        // Add each fetched record to the appropriate place
-        foreach ($fetchedRecords as $fetched) {
-            if (isset($addTo[$fetched->ID])) {
-                foreach ($addTo[$fetched->ID] as $addHere) {
-                    if ($addHere instanceof DataObject) {
-                        $addHere->setEagerLoadedData($relationName, $fetched);
-                    } elseif (is_array($addHere)) {
-                        $addHere['list']->addEagerLoadedData($relationName, $addHere['ID'], $fetched);
-                    } else {
-                        $this->eagerLoadedData[$relationChain][$addHere][$relationName] = $fetched;
+        foreach ($fetchedIDs as $class => $ids) {
+            foreach (DataObject::get($class)->byIDs($ids) as $fetched) {
```

Comment:
This change increases the number of requests that are made - and that number is increased regardless of whether the has_one is polymorphic or not.

Please refactor this to be a single query like it was before (instead of a query in a loop) or if that's not possible, only do it in a loop like this for polymorphic has_one.

Reply:
No its not. The `$fetchedIDs` will only contain one `$class` entry for a normal has_one, but multiple for polymorph.

Its does not make sense to do a if which checks if fetchedIDs only contains one entry todo the same thing which it would do in the loop ...

Reply:
Can you please try explaining that again? I don't understand what you're trying to say.

Reply:
With a normal `has_one` (`private static $has_one = ['Foo' => Foo::class];`), the `$fetchedIDs` array will only contain one entry (`Foo::class`). So only query to the Database will happen, like before this change.

With a polymorphic has_one (aka. `private static $has_one = ['Foo' => DataObject::class];`) you have following database schema:

| ID | FooClass | FooID |
| --- | ------------  | -------- |
| 1 | Member | 1 |
| 2 | Member | 2 |
| 3 | Group | 1 |

To  fetch these, you need a query to Member and one to Group table.

Reply:
Ahh, yup, that makes sense. Thank you for explaining that.